File: algorithms/data-structures/lru-cache/python/lru_cache.py
class _Node:
    __slots__ = ("key", "value", "prev", "next")

    def __init__(self, key: int = 0, value: int = 0) -> None:
        self.key = key
        self.value = value
        self.prev: "_Node | None" = None
        self.next: "_Node | None" = None


class _LRUCache:
    def __init__(self, capacity: int) -> None:
        self.capacity = capacity
        self.map: dict[int, _Node] = {}
        self.head = _Node()
        self.tail = _Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node: _Node) -> None:
        node.prev.next = node.next
        node.next.prev = node.prev

    def _add_to_head(self, node: _Node) -> None:
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node

    def get(self, key: int) -> int:
        if key in self.map:
            node = self.map[key]
            self._remove(node)
            self._add_to_head(node)
            return node.value
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.map:
            node = self.map[key]
            node.value = value
            self._remove(node)
            self._add_to_head(node)
        else:
            if len(self.map) == self.capacity:
                lru = self.tail.prev
                self._remove(lru)
                del self.map[lru.key]
            node = _Node(key, value)
            self.map[key] = node
            self._add_to_head(node)


def lru_cache(operations: list[int]) -> int:
    capacity = operations[0]
    op_count = operations[1]
    cache = _LRUCache(capacity)
    result_sum = 0
    idx = 2

    for _ in range(op_count):
        op_type = operations[idx]
        key = operations[idx + 1]
        value = operations[idx + 2]
        idx += 3

        if op_type == 1:
            cache.put(key, value)
        elif op_type == 2:
            result_sum += cache.get(key)

    return result_sum

File: algorithms/data-structures/lru-cache/kotlin/LruCache.kt
private class LruNode(var key: Int, var value: Int) {
    var prev: LruNode? = null
    var next: LruNode? = null
}

private class LruCacheImpl(private val capacity: Int) {
    private val map = HashMap<Int, LruNode>()
    private val head = LruNode(0, 0)
    private val tail = LruNode(0, 0)

    init {
        head.next = tail
        tail.prev = head
    }

    private fun removeNode(node: LruNode) {
        node.prev!!.next = node.next
        node.next!!.prev = node.prev
    }

    private fun addToHead(node: LruNode) {
        node.next = head.next
        node.prev = head
        head.next!!.prev = node
        head.next = node
    }

    fun get(key: Int): Int {
        val node = map[key] ?: return -1
        removeNode(node)
        addToHead(node)
        return node.value
    }

    fun put(key: Int, value: Int) {
        val existing = map[key]
        if (existing != null) {
            existing.value = value
            removeNode(existing)
            addToHead(existing)
        } else {
            if (map.size == capacity) {
                val lru = tail.prev!!
                removeNode(lru)
                map.remove(lru.key)
            }
            val node = LruNode(key, value)
            map[key] = node
            addToHead(node)
        }
    }
}

fun lruCache(operations: IntArray): Int {
    val capacity = operations[0]
    val opCount = operations[1]
    val cache = LruCacheImpl(capacity)
    var resultSum = 0
    var idx = 2

    for (i in 0 until opCount) {
        val opType = operations[idx]
        val key = operations[idx + 1]
        val value = operations[idx + 2]
        idx += 3

        if (opType == 1) {
            cache.put(key, value)
        } else if (opType == 2) {
            resultSum += cache.get(key)
        }
    }

    return resultSum
}
